### PR TITLE
Make DragSource extension/component available in DropEvent

### LINKED ISSUE #8439: Make DragSource extension/component available in DropEvent
When the drag has originated from the same UI, it would be nice to get the drag source component/extension in the drop event. This would make it really easy to pass the DnD operation data without using `dataTransfer.data` object when the DnD is happening inside the UI.

The easiest way to implement this might be to store it in the UI when a dragStart event has happened.

API:
```java
class DropEvent extends Component.Event {
...
  /**
   * If the drag had originated from this UI, this returns the source component.
   */
  public Optional<Component> getDragSourceComponent() {...}

  /**
   * If the drag had originated from this UI, this returns the source extension.
   */
  public Optional<DragSourceExtension> getDragSourceExtension() {...}
}
```

Could be easily done as a variable stored in the UI, that has public API that is used when the `DropEvent` happens:
```java
public void setActiveDragSource(DragSourceExtension extension) {...}
public DragSourceExtension getActiveDragSource() {...}
```
This needs to be cleared by the `DragSourceExtension` on a `dragend` event.

### PR DESCRIPTION
**Pull request by wbadam**
This will resolve #8439

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/framework/8636)

### CODE REVIEW

**Comment by pleku** [comment]
Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 189 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmKSK4r2g6Pe0jLk5:-KdZmKSK4r2g6Pe0jLk6:b-f192ys) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L189)):*
> ```Java
>             getRpcProxy(DropTargetRpc.class)
>                     .drop(types, data, getState().dropEffect,
>                             nativeEvent.getDataTransfer().getData(
> ```

What happens here if this data is not available, like when dragging from outside the application ?

---

*[server/src/main/java/com/vaadin/event/dnd/DropEvent.java, line 91 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmkzoGcYu8g6Mq2BS:-KdZmkzoGcYu8g6Mq2BT:bhws1a6) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/server/src/main/java/com/vaadin/event/dnd/DropEvent.java#L91)):*
> ```Java
>     /**
>      * Get the drag source component that has just been dropped if the drag
>      * source and drop target are in the same UI.
> ```

I'd rephrase to:
```
Returns the drag source component, if the drag originated from a component in the same UI as the drop target component, or an empty optional.

@ drag source component, or an empty optional

```

---

*[server/src/main/java/com/vaadin/event/dnd/DropEvent.java, line 96 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZoOQ9-kM9rux_anbl:-KdZoOQ9-kM9rux_anbm:b-hoa8op) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/server/src/main/java/com/vaadin/event/dnd/DropEvent.java#L96)):*
> ```Java
>      */
>     public Optional<T> getDragSourceComponent() {
>         return getDragSourceExtension().map(DragSourceExtension::getParent);
> ```

I think you should capture this during the event constructor call, because if there are several listeners for drop event, and some of them removes the drag source extension from the component, then this will be `null`

---

*[server/src/main/java/com/vaadin/event/dnd/DropEvent.java, line 101 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZo4JLsrjxtBGvDNNX:-KdZo4JLsrjxtBGvDNNY:b-jzzelu) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/server/src/main/java/com/vaadin/event/dnd/DropEvent.java#L101)):*
> ```Java
>     /**
>      * Get the drag source extension of the component that has just been dropped
>      * if the drag source and drop target are in the same UI.
> ```

Same type of rephrasing as above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8636)*

**Comment by wbadam** [comment]
Review status: 9 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 189 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmKSK4r2g6Pe0jLk5:-Kd_0VAqQXh1o-pbzTI-:b-79foxn) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L189)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

What happens here if this data is not available, like when dragging from outside the application ?
</blockquote></details>

According to specs it returns an empty string. Then the `getUI().getConnectorTracker().getConnector("")` would hopefully return `null`.

Should be checked explicitly if `DataTransfer` contains the type?

---

*[server/src/main/java/com/vaadin/event/dnd/DropEvent.java, line 91 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmkzoGcYu8g6Mq2BS:-Kd_3QAPAuHRPimWeIZi:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/server/src/main/java/com/vaadin/event/dnd/DropEvent.java#L91)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

I'd rephrase to:
```
Returns the drag source component, if the drag originated from a component in the same UI as the drop target component, or an empty optional.

@ drag source component, or an empty optional

```
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/DropEvent.java, line 96 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZoOQ9-kM9rux_anbl:-Kd_MBq_yU1Ax-yMY4GS:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/server/src/main/java/com/vaadin/event/dnd/DropEvent.java#L96)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

I think you should capture this during the event constructor call, because if there are several listeners for drop event, and some of them removes the drag source extension from the component, then this will be `null`
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/DropEvent.java, line 101 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZo4JLsrjxtBGvDNNX:-Kd_3QzRTdsvI4GRstf-:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/server/src/main/java/com/vaadin/event/dnd/DropEvent.java#L101)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Same type of rephrasing as above
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8636)*

**Comment by pleku** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 189 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmKSK4r2g6Pe0jLk5:-Kd_psVGTu83b1edSWWR:bdanu8q) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L189)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

According to specs it returns an empty string. Then the `getUI().getConnectorTracker().getConnector("")` would hopefully return `null`.

Should be checked explicitly if `DataTransfer` contains the type?
</blockquote></details>

Might be safe if some browser does not play nice

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8636)*

**Comment by wbadam** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 189 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmKSK4r2g6Pe0jLk5:-Kd_rPmMaySAkpS3b7ZV:b-kvmvw0) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L189)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Might be safe if some browser does not play nice
</blockquote></details>

Just realised that the data is already extracted from DataTransfer so the data map can simply be used.

Okay, last question. Should the dataSourceId appear as data in the server side event? Currently it does.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8636)*

**Comment by pleku** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 189 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmKSK4r2g6Pe0jLk5:-Kd_vVG9bZBne2o-juMi:b2cusxr) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L189)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

Just realised that the data is already extracted from DataTransfer so the data map can simply be used.

Okay, last question. Should the dataSourceId appear as data in the server side event? Currently it does.
</blockquote></details>

I think we can live with that for now, since it is a map people should anyways be using keys for that

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8636)*

**Comment by wbadam** [comment]
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 189 at r1](https://reviewable.io:443/reviews/vaadin/framework/8636#-KdZmKSK4r2g6Pe0jLk5:-Kd_zxUnsRaNoP2n6xqx:b-fc79xe) ([raw file](https://github.com/vaadin/framework/blob/e0033dbfa0611594847ff40830dba35b472e351c/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L189)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

I think we can live with that for now, since it is a map people should anyways be using keys for that
</blockquote></details>

Okay, then I'm done for now.

The data map is filled by iterating through the types getting the data from DataTransfer. If the data source ID is not present, data.get() will return null since it is a Java map (on the client but you get my point).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8636)*
